nilearn/decomposition/base_msdl.py

```diff
@@ -0,0 +1,511 @@
+            prev = selected
+            selected = []
+            for i in range(n_shuffle_subjects):
+                r = None
+                while r in selected or r in prev or r is None:
+                    r = np.random.randint(n_group)
+                ys.append(Ys[r])
+                us.append(Us[r])
+                selected.append(r)
+                unseen[r] = False
+            if verbose > 1:
+                print("Subjects selected: ", str(selected))
+            # We attribute group maps to unseen subjects
+            Vs[unseen] = V
+        else:
+            ys = Ys
+            us = Us
+
+        t0_ = time.time()
+        VsUs = Parallel(n_jobs=n_jobs)(delayed(_update_Vs_Us)
+                (ys_, us_, V, mu, verbose=verbose)
+                for (ys_, us_) in zip(ys, us))
+
+        t0 += time.time() - t0_
+        dt = t0
+
+        Vs_, residuals_dV_, Us_, r2_ = zip(*VsUs)
+
+        if n_shuffle_subjects is not None:
+            #dVs = 0.
+            for j, i in enumerate(selected):
+                Us[i] = Us_[j]
+                #dVs += np.sum(((Vs_[j] - Vs[i]) / (n_group * mu * 2.)) ** 2)
+                Vs[i] = Vs_[j]
+                residuals_dV[i] = residuals_dV_[j]
+                r2[i] = r2_[j]
+        else:
+            Vs_ = np.asarray(Vs_)
+            # Do that inplace to save memory
+            #Vs -= Vs_
+            #dVs = np.sum((Vs / (n_group * mu * 2.)) ** 2)
+            Vs = Vs_
+            Us = Us_
+            residuals_dV = residuals_dV_
+            r2 = r2_
+
+        if verbose > 2:
+            #print "Delta Vs: %f" % dVs
+            pass
+
+        current_cost = cost_function()
+        E.append(current_cost)
+        if verbose == 1:
+            sys.stdout.write('.')
+            sys.stdout.flush()
+        elif verbose:
+            print("Iteration % 3i.2 "
+                "(elapsed time: % 3is, % 4.1fmn, current cost % 7.3f)" %
+                    (ii, dt, dt / 60, current_cost))
+
+        dE = E[-2] - E[-1]
+        # We tolerate small increases in energy due to numerical
+        # error, but not much
+        #assert(dE > -tol * abs(E[-1]))
+        if dE < 0:
+            if verbose == 1:
+                sys.stdout.write('!')
+            elif verbose:
+                print('Energy increased by %.2f* tol' % (-dE / (tol * E[-1])))
+        elif last_iteration or dE < tol * abs(E[-1]):
+            if n_shuffle_subjects is not None:
+                last_iteration = True
+                n_shuffle_subjects = None
+                continue
+            if verbose == 1:
+                print("")
+            elif verbose:
+                print("--- Convergence reached after %d iterations" % ii)
+            if ii >= minit:
+                break
+        if ii % 5 == 0 and callback is not None:
+            callback(locals())
+
+    return Us, Vs, V, E
+
+
+# Gave me negative results on big dataset...
+def estimate_group_regularization(data, n_components,
```

Comment:
We should probably double check this one. In the real world, this should be very useful.